pilot/pkg/xds/endpoints/ep_filters.go

```diff
@@ -134,25 +182,70 @@ func (b *EndpointBuilder) EndpointsByNetworkFilter(endpoints []*LocalityEndpoint
 				Labels: labelutil.AugmentLabels(nil, gw.Cluster, "", "", gw.Network),
 			}
 
-			// Generate the EDS endpoint for this gateway.
-			gwEp := &endpoint.LbEndpoint{
-				HostIdentifier: &endpoint.LbEndpoint_Endpoint{
-					Endpoint: &endpoint.Endpoint{
-						Address: epAddr,
+			// Here we generate the EDS endpoint for the E/W gateways that replace the original endpoint.
+			// Depending on whether we operate in ambient mode or not, we generated endpoints for E/W
+			// gateways differently as we use somewhat different protocols in those two distinct cases.
+			var gwEp *endpoint.LbEndpoint
+
+			if isAmbientWaypoint {
+				gwAddr := gw.Addr
+				gwPort := int(gw.HBONEPort)
+
+				addr := net.JoinHostPort(gwAddr, strconv.Itoa(gwPort))
+				svcPort := b.servicePort(b.port)
+
+				gwEp = &endpoint.LbEndpoint{
+					HostIdentifier: &endpoint.LbEndpoint_Endpoint{
+						Endpoint: &endpoint.Endpoint{
+							// We need to redirect to an internal listener that will tunnel the data through
+							// a double-HBONE, we still use the E/W gateway address though for the endpoint
+							// id.
+							Address: util.BuildInternalAddressWithIdentifier(innerConnectOriginate, addr),
+						},
 					},
-				},
-				LoadBalancingWeight: &wrappers.UInt32Value{
-					Value: epWeight,
-				},
-				Metadata: &core.Metadata{},
+					LoadBalancingWeight: &wrappers.UInt32Value{
+						Value: epWeight,
+					},
+					Metadata: &core.Metadata{},
+				}
+
+				// TODO: figure out a way to extract locality data from the gateway public endpoints in meshNetworks
+				util.AppendLbEndpointMetadata(&model.EndpointMetadata{
+					Network: gw.Network,
+					// I don't think that TLSMode affects anythig downstream of this code anymore, but for ambient
+					// mode we do not rely on the legacy Istio mTLS, so I explicitly mark it as disabled.
+					TLSMode:   model.DisabledTLSModeLabel,
```

Comment:
Disabled likely has a different semantic if e.g. the proxy is sending to something not in the mesh. Should we create a new TLS mode for double hbone (or just use an hbone tls mode)?

Reply:
That's a fair point on mTLS mode and HBONE being ~distinct in the data model. If we do bring that context here it would probably be better as a separate field. I'm fine with this as is